atef/check.py

```diff
@@ -308,9 +465,39 @@ async def get_data_for_signal_async(
             executor=executor,
         )
 
+    def prepare(self, cache: Optional[DataCache] = None) -> None:
+        """
+        Implement in subclass to grab and cache dynamic values.
+        This is expected to set self.is_prepared to True if
+        successful.
+        """
+        # TODO: think about renaming this method, collides with PreparedComparison
```

Comment:
Beyond naming, it does _feel_ rather awkward when done at this layer.

Without digging into the implementation, I naively would assume that `Comparison` is fairly static (describing the configuration), then `PreparedComparison` wraps the `Comparison` to ready it for dynamic things - connecting to PVs, getting values from the control system - until it can finally execute the comparison and make its judgment. I wouldn't really have expected the signal cache to enter the equation at all in `atef.check`, for example. (I may have even contributed toward messing up the abstraction, so I'm in no way pointing fingers here...)

Regardless, the way it's implemented **works** most importantly!

Reply:
It was a point of contention in the previous iteration, for sure.  It is a sort of break in the axioms of atef, or at least the `Normal` - `Prepared` division.

If we were to move the signal-getting into the `PreparedComparison` constructors, those `PreparedComparison`'s would have to know how to handle those unique `DynamicValue`'s, taking that method away from the `Comparison`.  What does the best way of doing that look like?  `isinstance` checks for `HappiValue` and `EpicsValue`?  Another generation of `PreparedComparison` children specific to each `DynamicValue`?   That also felt weird to me, since we're moving the access logic away from the class that holds the access configuration information.  

I'm not super jazzed about it either, but it does work!  It has the interesting side-effect of showing the observed dynamic value in the configuration, as an example of sorts

Reply:
> those PreparedComparison's would have to know how to handle those unique DynamicValue's

You're saying that since different comparisons may have different dynamic values applied to various parts of the configuration, the only place this can really be done is here? E.g., otherwise the prepared comparison machinery would have to know which static attributes can be mapped to from each dynamic attribute?

It seems like the only alternative would be to have some other method that lets the prepared comparisons know the mapping from dynamic value to static value to resolve the dynamic behavior themselves, which might be needlessly indirect but would keep all the cache checking in one place.

(as a bonus, such a method could get some mileage in the gui portion of the code too)

Reply:
> You're saying that since different comparisons may have different dynamic values applied to various parts of the configuration, the only place this can really be done is here? E.g., otherwise the prepared comparison machinery would have to know which static attributes can be mapped to from each dynamic attribute?

Yes, and we'd end up spawning a new `isinstance` if-else tree, which I'd love to avoid

> It seems like the only alternative would be to have some other method that lets the prepared comparisons know the mapping from dynamic value to static value to resolve the dynamic behavior themselves, which might be needlessly indirect but would keep all the cache checking in one place.

This is a pattern that's been cropping up throughout atef, I think, and it's starting to feel bad from a maintainability point of view.  Every time we add something new, we have to make sure those all-knowing functions get updated as well.  (it's what happened in `atef.widgets.config.page.setup_multi_mode_for_widget`, if I understand your meaning)

Tangentially, the presence of these have made me worry about the overhead associated with adding new steps/comparisons.  If only it were as easy as just defining a new class and making a new widget.  Maybe as I continue to work on atef a cleaner solution will come to me...